x-pack/plugins/files/public/components/image/image.tsx

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import React, { MutableRefObject, useState, useRef, useEffect } from 'react';
+import type { FunctionComponent, ImgHTMLAttributes, LegacyRef } from 'react';
+import { Subscription } from 'rxjs';
+import { createViewportObserver } from './viewport_observer';
+
+export interface Props extends ImgHTMLAttributes<HTMLImageElement> {
+  src: string;
+  alt: string;
+  /**
+   * Emits when the image first becomes visible
+   */
+  onFirstVisible?: () => void;
+  ref?: LegacyRef<HTMLImageElement>;
+}
+
+/**
+ * A viewport-aware component that displays an image. This component is a very
+ * thin wrapper around the img tag.
+ *
+ * @note Intended to be used with files like:
+ *
+ * ```ts
+ * <Image src={file.getDownloadSrc(file)} ... />
+ * ```
+ */
+export const Image: FunctionComponent<Props> = ({ src, alt, ref, onFirstVisible, ...rest }) => {
+  const [isVisible, setIsVisible] = useState(false);
+  const [viewportObserver] = useState(() => createViewportObserver());
+  const subscriptionRef = useRef<undefined | Subscription>();
+  useEffect(() => () => subscriptionRef.current?.unsubscribe(), []);
+  return (
```

Comment:
Can we de-couple viewport observer from the `<img>` component? Into something like a render-prop:

```tsx
<ViewportSensor onVisible={() => {}}>
  {(isVisible, ref) => <Image {...props} ref={ref} visible={isVisible} blurHash={'xyz...'} />}
<ViewportSensor>
```

Reply:
In the end I factored it out to a separate hook, can easily be converted to a child props API too if need be in future 👍🏻